include/seastar/coroutine/async_generator.hh

```diff
@@ -0,0 +1,421 @@
+                      std::coroutine_handle<> consumer) noexcept
+            : _promise{promise}
+            , _consumer{consumer}
+        {}
+        bool await_ready() const noexcept {
+            return false;
+        }
+        template <typename Promise>
+        std::coroutine_handle<> await_suspend(std::coroutine_handle<Promise> producer) noexcept {
+            _promise->_waiting_task = &producer.promise();
+            return _consumer;
+        }
+        void await_resume() noexcept {}
+    };
+
+    yield_awaiter do_yield() noexcept {
+        return yield_awaiter{this, _consumer};
+    }
+
+    class copy_awaiter {
+        using value_type = std::remove_cvref_t<Yielded>;
+        async_generator_promise_base* _promise;
+        std::coroutine_handle<> _consumer;
+        value_type _value;
+    public:
+        copy_awaiter(async_generator_promise_base* promise,
+                     std::coroutine_handle<> consumer,
+                     const value_type& value)
+            : _promise{promise}
+            , _consumer{consumer}
+            , _value{value}
+        {}
+        constexpr bool await_ready() const noexcept {
+            return false;
+        }
+        template <typename Promise>
+        std::coroutine_handle<> await_suspend(std::coroutine_handle<Promise> producer) noexcept {
+            auto& current = producer.promise();
+            current._value = std::addressof(_value);
+            _promise->_waiting_task = &current;
+            return _consumer;
+        }
+        constexpr void await_resume() const noexcept {}
+    };
+
+public:
+    async_generator_promise_base() noexcept = default;
+    async_generator_promise_base(const async_generator_promise_base &) = delete;
+    async_generator_promise_base& operator=(const async_generator_promise_base &) = delete;
+    async_generator_promise_base(async_generator_promise_base &&) noexcept = default;
+    async_generator_promise_base& operator=(async_generator_promise_base &&) noexcept = default;
+
+    // lazily-started coroutine, do not execute the coroutine until
+    // the coroutine is awaited.
+    std::suspend_always initial_suspend() const noexcept {
+        return {};
+    }
+
+    yield_awaiter final_suspend() noexcept {
+        _value = nullptr;
+        return yield_awaiter{this, this->_consumer};
+    }
+
+    void unhandled_exception() noexcept {
+        _exception = std::current_exception();
+    }
+
+    yield_awaiter yield_value(Yielded value) noexcept {
+        this->_value = std::addressof(value);
+        return yield_awaiter{this, this->_consumer};
+    }
+
+    copy_awaiter yield_value(const std::remove_reference_t<Yielded>& value)
+    requires (std::is_rvalue_reference_v<Yielded> &&
+              std::constructible_from<
+                  std::remove_cvref_t<Yielded>,
+                  const std::remove_reference_t<Yielded>&>) {
+        return copy_awaiter{this, this->_consumer, value};
+    }
+
```

Comment:
I don't understand this maze of awaiters. I need some documentation that maps user syntax (co_yield and co_await calls) to awaiters and some theory of operation. This stuff is too complicated to reverse engineer.

Reply:
i understand your perspective. i share similar sentiments while working on this. to address this, I will to create accompanying documentation specifically targeting developers to explain the reason why we need them, and how we address the problem.